src/engine/sparqlExpressions/ConvertToDtypeConstructor.cpp

```diff
@@ -130,44 +130,34 @@ namespace detail::to_datetime {
 
 // Cast to xsd:dateTime or xsd:date (ValueId)
 template <bool ToJustXsdDate>
-inline auto convertStringToDateTimeValueId =
-    [](OptIdOrString input) -> ValueId {
-  if (!input.has_value()) {
-    return Id::makeUndefined();
-  }
-  const auto& inputValue = input.value();
-
-  // Remark: If the parsing procedure for datetime/date string values with
-  // parseXsdDatetimeGetOptDate/parseXsdDateGetOptDate fails,
-  // return Id::makeUndefined().
-  const auto retrieveValueId = [](std::optional<DateYearOrDuration> optValue) {
-    if (optValue.has_value()) {
-      return Id::makeFromDate(optValue.value());
-    }
-    return Id::makeUndefined();
-  };
-
-  if (auto* valueId = std::get_if<ValueId>(&inputValue)) {
-    return *valueId;
-  }
-
-  auto* str = std::get_if<std::string>(&inputValue);
-  AD_CORRECTNESS_CHECK(str != nullptr);
-  if constexpr (ToJustXsdDate) {
-    return retrieveValueId(DateYearOrDuration::parseXsdDateGetOptDate(*str));
-  } else {
-    return retrieveValueId(
-        DateYearOrDuration::parseXsdDatetimeGetOptDate(*str));
-  }
+inline const auto castStringToDateTimeValueId = [](OptIdOrString input) {
+  if (!input.has_value()) return Id::makeUndefined();
+
+  using DYD = DateYearOrDuration;
+  std::optional<DYD> optValueId = std::visit(
+      [&]<typename T>(const T& value) {
+        if constexpr (std::is_same_v<std::decay_t<T>, ValueId>) {
+          return ToJustXsdDate ? DYD::convertToXsdDate(value.getDate())
```

Comment:
add a CORRECTNESS CHECK that the ID in fact contains a date.
Or, even better:
Let the value getter handle this (make it a "StringOrDateGetter", that's the correct thing to do hre.